velero-csi-plugin/csi_snapshotter.go

```diff
@@ -117,10 +140,40 @@ func (p *CSISnapshotter) Execute(item runtime.Unstructured, backup *v1.Backup) (
 	if annotations == nil {
 		annotations = make(map[string]string)
 	}
-	annotations["velero.io/volume-snapshot-name"] = upd.Name
+	annotations[veleroSnapshotNameAnnotation] = upd.Name
+	annotations[veleroBackupNameAnnotation] = backup.Name
 	accessor.SetAnnotations(annotations)
 
-	return item, nil, nil
+	//TODO: add retry logic here. CSI driver may take time to reconcile the snapshotcontents object
+	snapshotContents, err := getVolumeSnapshotContentForVolumeSnapshot(snapshotClient.SnapshotV1beta1(), upd.Name)
```

Comment:
Does this mean that we're pausing until the contents are completely ready?

I don't think we need to do that, do we? We simply need the pointer to the contents object, so we know what to fetch and include.

Reply:
The retry logic here is for the CSI driver to create the snapshotcontent object. This querying of the snapshotcontent object can occur before the CSI driver has had a change to create the snapshotcontent object.

Also, I think we should wait for `VolumeSnapshot.Status.readyToUse` to be true. If we don't, then it is possible that the velero backup process, while backing up the volumesnapshot as the additional items, can make progress without the CSI driver having completed its snapshotting. Leading to the possibility that we backup a Volumesnapshot that is not ready to be use, i.e., not ready to be restored. Which is what may be causing https://github.com/vmware-tanzu/velero-plugin-for-csi/issues/1

Reply:
> Also, I think we should wait for VolumeSnapshot.Status.readyToUse to be true.

We can discuss this in a separate issue, as I don't want to lump all work into this PR.

However, I will say that the design of CSI is such that the intention is _not_ to pause all backup work until `readyToUse == true`. This is also similar to how Velero currently works, by launching an async request against the storage provider's snapshot APIs, then proceeding.

While I'm aware that this can get into the state that it's not ready to restore, the restoration state should be reconciled when the `VolumeSnapshot` and `VolumeSnapshotContent` objects are recreated in the new cluster. Their `Status` fields will be cleared upon restoration by Velero, and the on-disk snapshot itself will be queried for readiness. This would be considered [an import case](https://kubernetes.io/blog/2018/10/09/introducing-volume-snapshot-alpha-for-kubernetes/#importing-an-existing-snapshot-with-kubernetes) in the upstream terminology.

Reply:
Yeah, we can discuss whether or not we should wait for `VolumeSnapshot.Status.readyToUse` in a separate issue.

Reply:
This is an existing issue with Velero & non-CSI snapshots too -- see https://github.com/vmware-tanzu/velero/issues/1799 - so I'd like to think about a solution for both.